### PR TITLE
Feature/rest api tests

### PR DESCRIPTION
**Pull request by MazanYan**
* Tests for POST /users/signup
* Tests for GET /r/{alias}
* Tests for GET /urls
* Tests for DELETE /urls/{alias}
* Switched to JsonToolJacksonImpl both in REST API and tests
* Added missing UserAlreadyExistsExceptionHandler
* More descriptive messages for SignUp exceptions

### CODE REVIEW

File: src/main/java/edu/kpi/testcourse/rest/UrlController.java [hunk 1]
```diff
@@ -35,6 +35,9 @@
   @Inject
   AuthorizationMockServiceImpl authorizationMockService;
 
+  @Inject
+  JsonToolJacksonImpl jsonTool;
```
Comment (RingaVirda) [inline src/main/java/edu/kpi/testcourse/rest/UrlController.java:39]: You should inject JsonTool, JsonToolJacksonImpl will be chosen as only implementation now existent, or you can configure it in BeanFactory.
```suggestion
  JsonTool jsonTool;
```

File: src/main/java/edu/kpi/testcourse/rest/UrlController.java [hunk 2]
```diff
@@ -94,8 +97,10 @@ public String shortenUrl(@Header("token") String token, @Body ShortLink urlToSho
    * @return list of urls
    */
   @Get()
-  public MutableHttpResponse<LinksOfUser> getUserUrls(@Header String token) {
+  public MutableHttpResponse<String> getUserUrls(@Header String token) {
     String email = this.authorizationMockService.authorizeUser(token);
-    return HttpResponse.ok(new LinksOfUser(this.shortLinkService.getLinksByUserEmail(email)));
+    LinksOfUser linksOfUser = new LinksOfUser(this.shortLinkService.getLinksByUserEmail(email));
```
Comment (RingaVirda) [inline src/main/java/edu/kpi/testcourse/rest/UrlController.java:102]: Is it necessary to use LinksOfUser? Wrapping object doesn't do anything here, you could response with raw list.

File: src/main/java/edu/kpi/testcourse/rest/UserController.java [hunk 3]
```diff
@@ -21,8 +19,8 @@
   private UserService userService;
 
   @Post(value = "/signup")
-  public HttpResponse<?> signup(@Body User user) throws SignUpException {
+  public HttpResponse<String> signup(@Body User user) throws SignUpException {
     userService.signup(user);
-    return HttpResponse.ok();
+    return HttpResponse.created("");
```
Comment (RingaVirda) [inline src/main/java/edu/kpi/testcourse/rest/UserController.java:24]: Is it ok to response with 201 here? It was written in specification to create Ok response on this endpoint.

File: src/test/java/edu/kpi/testcourse/rest/RedirectLinkTest.java [hunk 4]
```diff
@@ -86,45 +107,85 @@ public static void stopServer() {
...
-        g.toJson(
-          new ShortLink(
-            "haskell2",
-            "user1@mail.com",
-            "https://darcs.realworldhaskell.org/static/00book.pdf"
-          )
-        )
-      ).header("token", TEST_VALID_TOKEN)
+    // Micronaut seems to have different approach in testing redirect responses
+    // then the Publisher/Subscriber contract is used in the test
```
Comment (RingaVirda) [inline src/test/java/edu/kpi/testcourse/rest/RedirectLinkTest.java:111]: Yeah very unintuitive code

File: src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java [hunk 5]
```diff
@@ -27,7 +31,22 @@
 
   private static EmbeddedServer server;
   private static HttpClient client;
-  private static final Gson g = new Gson();
+
+  @Inject
+  JsonToolJacksonImpl jsonTool;
```
Comment (RingaVirda) [inline src/test/java/edu/kpi/testcourse/rest/ShortenLinkTest.java:36]: The same as in UrlController
```suggestion
  JsonTool jsonTool;
```

File: src/test/java/edu/kpi/testcourse/rest/SignUpTest.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
+import javax.inject.Inject;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+@MicronautTest
+public class SignUpTest {
+  @Inject
+  JsonToolJacksonImpl jsonTool;
```
Comment (RingaVirda) [inline src/test/java/edu/kpi/testcourse/rest/SignUpTest.java:22]:
```suggestion
  JsonTool jsonTool;
```

**Review by RingaVirda — COMMENTED** [review]
Very very scary tests

File: src/main/java/edu/kpi/testcourse/rest/UrlController.java (hunk 2 above)
Comment (MazanYan) [reply src/main/java/edu/kpi/testcourse/rest/UrlController.java:102]: I cannot respond with a raw list.
The structure of JSON response:
```
{
    "urls": [
        ...linksOfUser
    ]
}
```
I could wrap the ArrayList into `Collections.singletonMap("urls", this.shortLinkService.getLinksByUserEmail(email))`, but the separate DTO seems to be a more clear approach in my opinion.

File: src/main/java/edu/kpi/testcourse/rest/UserController.java (hunk 3 above)
Comment (MazanYan) [reply src/main/java/edu/kpi/testcourse/rest/UserController.java:24]: The basic specification tells us that the 201 status code is the correct response. Sorry about the mistake in the design document

Screenshot_20210318_103717

File: src/main/java/edu/kpi/testcourse/rest/UrlController.java (hunk 1 above)
Comment (MazanYan) [reply src/main/java/edu/kpi/testcourse/rest/UrlController.java:39]: Thanks for the good note! Checked your and Vadim's code - will fix my part of the project.

**Review by RingaVirda — APPROVED** [review]

**Pull request merged**